src/wasm.rs

```diff
@@ -74,44 +76,85 @@ impl Future for InnerFuture {
 }
 
 mod fetch {
+    use futures_util::io::AsyncReadExt;
     use js_sys::{Array, ArrayBuffer, Reflect, Uint8Array};
     use wasm_bindgen::JsCast;
     use wasm_bindgen_futures::JsFuture;
-    use web_sys::window;
-    use web_sys::RequestInit;
+    use web_sys::{window, RequestInit};
 
     use std::io;
     use std::iter::{IntoIterator, Iterator};
+    use std::pin::Pin;
 
     /// Create a new fetch request.
-    pub(crate) fn new(method: impl AsRef<str>, url: impl AsRef<str>) -> Request {
-        Request::new(method, url)
-    }
 
     /// An HTTP Fetch Request.
     pub(crate) struct Request {
-        init: RequestInit,
-        url: String,
+        request: web_sys::Request,
+        _body_buf: Pin<Vec<u8>>,
     }
 
     impl Request {
         /// Create a new instance.
-        pub(crate) fn new(method: impl AsRef<str>, url: impl AsRef<str>) -> Self {
-            let mut init = web_sys::RequestInit::new();
-            init.method(method.as_ref());
-            Self {
-                init,
-                url: url.as_ref().to_owned(),
+        pub(crate) async fn new(mut req: super::Request) -> Result<Self, io::Error> {
+            //create a fetch request initaliser
+            let mut init = RequestInit::new();
+
+            //set the fetch method
+            init.method(req.method().as_ref());
+
+            let uri = req.url().to_string();
+            let mut body = req.take_body();
+
+            //convert the body into a uint8 array
+            // needs to be pinned and retained inside the Request because the Uint8Array passed to
+            // js is just a portal into WASM linear memory, and if the underlying data is moved the
+            // js ref will become silently invalid
+            let mut body_buf = Vec::with_capacity(1024);
```

Comment:
Clever; I wasn't aware of this. Is this what was causing the bug?

Reply:
This is part of the fix for #3, originally from PR #20. I forgot to update the title and description when I rebased #20, so I didn't give you all the context for this fix. Sorry about that!